text/0048-session-keys-runtime-api.md

```diff
@@ -0,0 +1,102 @@
+# RFC-0048: Generate ownership proof for `SessionKeys`
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 13 November 2023                                                                    |
+| **Description** | Change `SessionKeys` runtime api to also create a proof of ownership for on chain registration. |
+| **Authors**     | Bastian Köcher                                                                    |
+
+## Summary
+
+When rotating/generating the `SessionKeys` of a node, the node calls into the runtime using the 
+`SessionKeys::generate_session_keys` runtime api. This runtime api function needs to be changed
+to add an extra parameter `owner` and to change the return value to also include the `proof` of 
+ownership. The `owner` should be the account id of the account setting the `SessionKeys` on chain 
+to allow the on chain logic the verification of the proof. The on chain logic is then able to proof 
+the possession of the private keys of the `SessionKeys` using the `proof`.
+
+## Motivation
+
+When a user sets new `SessionKeys` on chain the chain can currently not ensure that the user 
+actually has control over the private keys of the `SessionKeys`. With the RFC applied the chain is able 
+to ensure that the user actually is in possession of the private keys.
+
+## Stakeholders
+
+- Polkadot runtime implementors
+- Polkadot node implementors
+- Validator operators
+
+## Explanation
+
+We are first going to explain the `proof` format being used:
+```rust
+type Proof = (Signature, Signature, ..);
+```
+
+The `proof` being a SCALE encoded tuple over all signatures of each private session 
+key signing the `owner`. The actual type of each signature depends on the
+corresponding session key cryptographic algorithm. The order of the signatures in 
+the `proof` is the same as the order of the session keys in the `SessionKeys` type.
+
+The version of the `SessionKeys` needs to be bumped to `1` to reflect the changes to the 
+signature of `SessionKeys_generate_session_keys`:
+```rust
+pub struct OpaqueGeneratedSessionKeys {
+	pub keys: Vec<u8>,
+	pub proof: Vec<u8>,
+}
+
+fn SessionKeys_generate_session_keys(owner: Vec<u8>, seed: Option<Vec<u8>>) -> OpaqueGeneratedSessionKeys;
```

Comment:
An `AccountId` is a u64 I believe. How does that translate to a `Vec<u8>`?
Since the function signature here is the "high level one", shouldn't it be `u64` here?

Reply:
You'd presumably sign the actual controller key here, not some reference.

Reply:
> An `AccountId` is a u64 I believe. How does that translate to a `Vec<u8>`?

`AccountId` for Polkadot is an enum with public keys of sr/ed25519 or ecdsa. Other chain can also have different account ids. Thus, this should be a `Vec<u8>` that represents the SCALE encoded account id.